File: src/main/java/com/raychenon/leetcode/LongestSubstringWithoutRepeatingCharacters.java
package com.raychenon.leetcode;

import java.util.HashSet;
import java.util.Set;

/**
 * User: raychenon
 * Date: 2019-12-03
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */
public class LongestSubstringWithoutRepeatingCharacters {

    /**
     * time complexity: O(n).
     * space complexity: O(min(string_length,alphabet_count)),
     * if the string is longer than 26 (total of aplhabet letters), then the string repeats characters
     */
    public int lengthOfLongestSubstring(String s) {
        int start = 0, end = 0;
        int maxLen = 0;
        Set<Character> charsInCurrentSubStringSet = new HashSet<>();
        while (start < s.length() && end < s.length()) {
            if (!charsInCurrentSubStringSet.contains(s.charAt(end))) {
                charsInCurrentSubStringSet.add(s.charAt(end++));
                maxLen = Math.max(maxLen, end - start);
            } else {
                charsInCurrentSubStringSet.remove(s.charAt(start++));
            }
        }
        return maxLen;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/LongestSubstringWithoutRepeatingCharacters.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-12-03
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * Given a string, find the length of the longest substring without repeating characters.
 */
object LongestSubstringWithoutRepeatingCharacters {

    fun lengthOfLongestSubstring(s: String): Int {
        var start = 0
        var end = start

        var map = mutableMapOf<Char, Int>()

        var c = s[0]
        map.put(c, 1)
        var maxLen = 0
        for (i in 1..s.length - 1) {
            if (map.containsKey(s[i])) {
                end = i
                maxLen = Math.max(maxLen, end - start)
            }

            // equal to previous char
            if (c == s[i]) {
                start = i
            }

            map.put(s[i], map.getOrDefault(s[i], 0) + 1)
            c = s[i]
        }

        return maxLen

    }
}